ipc: Share /dev/shm and /dev/mqueue when --ipc container:<id/name> is used

mrunalp opened the pull request:
See #12046 

When container A has a user mounted /dev/shm and container B is started with the flag --ipc container:A, then the same /dev/shm volume is mounted into B.

Signed-off-by: Mrunal Patel mrunalp@gmail.com

cpuguy83 commented:
I don't like this.
Why don't we instead mount bind-mount a /dev/shm from the container config dir for all containers, and when someone uses `--ipc container:A` we use container A's instead of creating one?

mrunalp commented:
@cpuguy83 I don't follow what you are suggesting. Are you saying that /dev/shm should be bind mounted into A from a known location and then B can re-use it when it specified --ipc container:A?

By default /dev/shm is mounted within a containers namespace and we would want to retain that behavior. When A has a /dev/shm bind mounted in, then we know that we can share it with B otherwise it isn't possible.

thockin commented:
Why is this only about user-mounted?  I think it should always be true.

mrunalp commented:
@thockin That would mean supporting a bind mounted /dev/shm for each container by default or through a flag.

cpuguy83 commented:
@mrunalp Yes, that's what I'm advocating for, bind-mounted /dev/shm for each container, `--ipc container:A`, means use `A`'s instead of creating one for this new container.

mrunalp commented:
@cpuguy83 Do we want to do it for every container by default or have a flag enable it? There could be cases where containers would still want their own mount namespace private /dev/shm like today.

thockin commented:
I think that it should be governed by the IPC namespace, full stop.

On Fri, Apr 10, 2015 at 4:51 PM, Mrunal Patel <notifications@github.com>
wrote:

> @cpuguy83 <https://github.com/cpuguy83> Do we want to do it for every
> container by default or have a flag enable it? There could be cases where
> containers would still want their own mount namespace private /dev/shm like
> today.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/12159#issuecomment-91720611>.
>

mrunalp commented:
@thockin The question here is what happens when I start just a single container. Does it get a bind mounted /dev/shm (required for sharing with other containers) or it doesn't?

thockin commented:
oh.  Well, you can't know a priori, so yes :)  IMHO

On Fri, Apr 10, 2015 at 4:57 PM, Mrunal Patel <notifications@github.com>
wrote:

> @thockin <https://github.com/thockin> The question here is what happens
> when I start just a single container. Does it get a bind mounted /dev/shm
> (required for sharing with other containers) or it doesn't?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/12159#issuecomment-91722226>.
>

cpuguy83 commented:
Yep, would need to do it for all containers in this case.

mrunalp commented:
Should we mount the /dev/shm and /dev/mqueue for a container at /var/lib/container/<container_id>/dev/shm and then bind mount it in? Does that location sound good? If so, then I can start implementing that.

LK4D4 commented:
`/dev/shm` and `/dev/mqueue` mount in `/var/lib/docker/containers/...` sounds okay for me. I think we need one-two more approval.
ping @icecrime @crosbymichael @jfrazelle @tiborvass @calavera

crosbymichael commented:
This one is tricky.  Are there any numbers on this that show making this change improves things in a meaningful way?  Is it worth the cost of complexity?

mrunalp commented:
@crosbymichael A simpler alternative is this PR. If A has /dev/shm volume mounted in and B does --ipc container:B, then it shares the same /dev/shm.

cpuguy83 commented:
After a review with @icecrime @tiborvass @duglin @tianon 

We agree that with `--ipc container`, sharing should just be automatic.
The implementation on this PR is, however, not desirable as expressed in the comments above.
There shouldn't be any dependency on volumes.

Do you have any other suggestions on a design for this?

thockin commented:
Is there another way to share files across containers?

mrunalp commented:
@cpuguy83 Only way to make this work is to have a directory (with either shm or mqueue) on the host and mount it in each container. -v host:container already provides that.

icecrime commented:
Can't we "inject" it as a mount (i.e., `container.MountPoints`) rather than as a volume, something like `if ipcModeIsContainer() { container.addBindMountPoint(...) }`?

I don't like that this will appear in the `inspect` but wasn't really asked for.

mrunalp commented:
@icecrime We could do that. We do however need to create /dev/shm and /dev/mqueue mount points for each container on the host and bind mount them in. When it is shared we just re-use the mount of the target container.

icecrime commented:
@mrunalp Sounds good to me!

calavera commented:
@mrunalp I believe that's the right behavior. If you use `addBindMountPoint` for one container, the subsequent `addBindMountPoint` in other containers will re-use the mount of the target container.

rhatdan commented:
I think you should clean the disk so that it will work like tmpfs.  Must make sure that it works with SELinux, also this will give you worse performance, but not sure how much these directories are written too, since you would no longer be using an in memory file system.

tiborvass commented:
Ping @mrunalp do you have time to carry on with this? If not, let us know!

mrunalp commented:
@tiborvass: @rhatdan and I are working on it. Hopefully should have a PR ready this week.

tiborvass commented:
@mrunalp will you update this PR or should we close it?

mrunalp commented:
@tiborvass I just updated this PR.

cpuguy83 commented on daemon/container_unix.go:
```diff
@@ -1130,3 +1148,64 @@ func (container *Container) PrepareStorage() error {
+		return err
+	}
+
+	if err := syscall.Mount("shm", shmPath, "tmpfs", uintptr(syscall.MS_NOEXEC|syscall.MS_NOSUID|syscall.MS_NODEV), label.FormatMountLabel("mode=1777,size=65536k", container.GetMountLabel())); err != nil {
```
use the mount package?

cpuguy83 commented:
Sorry added code review comment though it's in design review.

Design is ok for me.

rhatdan commented:
One big problem is when to umount the content tmpfs on /var/lib/docker.  

Container1 runs 
container2 --ipc:container:container1
docker stop container1

Do we attempt to umount on container exit, it will fail.

docker stop container2
Do we attempt to umount again?

If I stop first container and attempt to delete it, do we handle this properly.

mrunalp commented:
>  One big problem is when to umount the content tmpfs on /var/lib/docker.

> Container1 runs
> container2 --ipc:container:container1
> docker stop container1

> Do we attempt to umount on container exit, it will fail.

We unmount shm/mqueue for container1 at this point.

> docker stop container2
> Do we attempt to umount again?

We do not attempt to umount as this container did not own those mounts.

> If I stop first container and attempt to delete it, do we handle this properly.

Yes, I see no errors.

rhatdan commented:
What does the second container see in /dev/shm if the first one disappears.

mrunalp commented:
@rhatdan It seems to inherit the contents in a private /dev/shm and we are able to read/write from it.

mrunalp commented:
@rhatdan I guess it isn't good as it hides the fact that the underlying bind has disappeared.

mrunalp commented:
@rhatdan I think it could be fixed by making it slave instead of private. As it is private, the umount from the host doesn't get propagated to it.

icecrime commented:
*Collective review session*

Thanks @mrunalp! Moving to code review as it seems that the new implementation fits the comments we had above.

If anybody seems remaining objections about this, please speak up.

rhatdan commented:
I figured out why removing container1 and umounting the original tmpfs does not effect container2.  Container2 creates a separate mnt namespace and inherits all of the mounts in the docker daemon mount namespace.  If the docker daemon later unmounts a file system, my mount namespace will still have a reference to the mount point, so it will stay mounted in my mount namespace, it will only disappear when my mount namespace goes away.

@jeremyeder do you think this could cause performance problems?

@rhvgoyal Thinks there would not be a performance hit.

mrunalp commented:
I added code to cleanup old shm/mqueue mounts on daemon startup.

jeremyeder commented:
This fails (this has always been my CMD, just added the --ipc)
```
# docker run --ipc=container:311ef95655b2 -d r7perf /usr/sbin/httpd -DFOREGROUND
```

These succeed:
```
# docker run --ipc=container:311ef95655b2 -d r7perf sleep 99999
# docker run --ipc=container:311ef95655b2 -it r7perf bash
```

Haven't seen this behavior before...the container definitely starts, executes httpd and then just exits immediately with no useful logging, even with docker -D.

So, I'm running the 4000 container --ipc test using 'sleep' for now...should be fine, being that we're testing mounts.  But the failure with httpd needs more investigation.

rhatdan commented:
@jeremyeder Any chance SELinux issue on httpd, httpd acually uses /dev/shm might be getting a permission denied and dieing

jeremyeder commented:
docker dumped core @ 2498 containers.

rhatdan commented:
@jeremyeder Do we know how many we can get without the patch?

jeremyeder commented:
That's running now.  Give me a half hour.

jeremyeder commented:
Unpatched docker failed at the same exact place.  Debugging now.

rhatdan commented:
Excellent, I like bug for bug compatibility.

jeremyeder commented:
OK, found the cause:

```
Jul 28 16:05:00 server docker: runtime: program exceeds 10000-thread limit
Jul 28 16:05:00 server docker: fatal error: thread exhaustion
Jul 28 16:05:00 server docker: runtime stack:
Jul 28 16:05:00 server docker: runtime.throw(0x129a3c1)
Jul 28 16:05:00 server docker: /usr/lib/golang/src/runtime/panic.go:491 +0xad fp=0x7fa56f709be8 sp=0x7fa56f709bb8
```

And there's about 3000 lines of backtrace after that.

This error occurred on docker-1.7.0-4.el7.x86_64 (which does not include this IPC patch) as well as upstream docker master experimental=true built last Friday.  So this thread limit has nothing to do with this patch. 

This is the last update about docker panics that I'll put into this issue...need to dig more into this.  The next update will be results of the 1000 container scaling test...Will try to have those results tomorrow.  Sorry for the detour.

vbatts commented:
@jeremyeder can you open a new issue with a reproducer i can use?

jeremyeder commented:
I ran a comparison using a docker build from https://github.com/mrunalp/docker.git feature/ipc_share_dev

A) 1000 rhel7 containers
B) 1000 rhel7 containers, all sharing the IPC namespace with a "control" container.

Verified I can read/write shared memory between containers.

After looking at the system and benchmark metrics, I don't see any change in performance or scalabillity.  :+1:

rhatdan commented:
Awesome

cpuguy83 commented on daemon/daemon.go:
```diff
@@ -745,9 +747,46 @@ func NewDaemon(config *Config, registryService *registry.Service) (daemon *Daemo
 }
 
+// cleanupMounts umounts shm/mqueue mounts for old containers
+func (daemon *Daemon) cleanupMounts() error {
```
This should go into a linux specific file.

cpuguy83 replied:
Also, shouldn't this run as part of daemon shutdown as well? I'd prefer to not leave mounts around.

mrunalp replied:
done.

cpuguy83 commented on daemon/daemon.go:
```diff
@@ -745,9 +747,46 @@ func NewDaemon(config *Config, registryService *registry.Service) (daemon *Daemo
+			mnt := fields[4]
+			mountBase := filepath.Base(mnt)
+			if mountBase == "mqueue" || mountBase == "shm" {
+				logrus.Infof("Unmounting %+v\n", mnt)
```
This is unnecessary... maybe debug.

mrunalp replied:
done.

cpuguy83 commented on daemon/daemon.go:
```diff
@@ -745,9 +747,46 @@ func NewDaemon(config *Config, registryService *registry.Service) (daemon *Daemo
 
+// cleanupMounts umounts shm/mqueue mounts for old containers
+func (daemon *Daemon) cleanupMounts() error {
+	logrus.Infof("Cleaning up old shm/mqueue mounts: start.")
```
These logs probably don't need to be Info level either.

mrunalp replied:
done.

mrunalp commented:
Thanks for the performance testing @jeremyeder :)

cpuguy83 commented on daemon/container_unix.go:
```diff
@@ -201,14 +202,31 @@ func populateCommand(c *Container, env []string) error {
+		return err
+	}
+
+	c.MqueuePath, err = c.GetRootResourcePath("mqueue")
```
`c.mqueuePath()`

mrunalp replied:
done.

cpuguy83 commented on daemon/container_unix.go:
```diff
@@ -201,14 +202,31 @@ func populateCommand(c *Container, env []string) error {
 	ipc := &execdriver.Ipc{}
 
+	var err error
+	c.ShmPath, err = c.GetRootResourcePath("shm")
```
`c.shmPath()`

mrunalp replied:
done.

cpuguy83 commented:
Tests would be awesome.

mrunalp committed:
Add support for sharing /dev/shm/ and /dev/mqueue between containers

This changeset creates /dev/shm and /dev/mqueue mounts for each container under
/var/lib/containers/<id>/ and bind mounts them into the container. When --ipc:container<id/name>
is used, then the /dev/shm and /dev/mqueue of the ipc container are used instead of creating
new ones for the container.

Signed-off-by: Mrunal Patel <mrunalp@gmail.com>
Docker-DCO-1.1-Signed-off-by: Dan Walsh <dwalsh@redhat.com> (github: rhatdan)

mrunalp commented:
Rebased.

icecrime commented:
@mrunalp I can't tell if it's related to your changes or a bad rebase, but we have repeated tests failure here :-(

mrunalp commented:
@icecrime The test that is failing in jenkins complains about shm/mqueue mounts left remaining when the docker daemon is killed. However, this PR includes the code to clean those mounts on startup. So, I think that we need to merge the clean up code and then update jenkins server's docker daemon to include that code to make the tests pass.

crosbymichael commented:
I got this locally 

FAIL: docker_cli_daemon_test.go:1414: DockerDaemonSuite.TestCleanupMountsAfterCrash

mrunalp commented:
I built an ubuntu vm to try to reproduce this and looks like that this test is passing.

```
root@ubuntu:~/gosrc/src/github.com/docker/docker# TESTFLAGS='-check.f DockerSuite.TestCleanupMountsAfterCrash' ./hack/make.sh binary test-integration-cli
# WARNING! I don't seem to be running in the Docker container.
# The result of this command might be an incorrect build, and will not be
#   officially supported.
#
# Try this instead: make all
#

bundles/1.8.0-dev already exists. Removing.

---> Making bundle: binary (in bundles/1.8.0-dev/binary)
Building: bundles/1.8.0-dev/binary/docker-1.8.0-dev
Created binary: bundles/1.8.0-dev/binary/docker-1.8.0-dev

---> Making bundle: test-integration-cli (in bundles/1.8.0-dev/test-integration-cli)
---> Making bundle: .integration-daemon-start (in bundles/1.8.0-dev/test-integration-cli)
+++ /etc/init.d/apparmor start
Starting apparmor (via systemctl): apparmor.service.
+++ /sbin/apparmor_parser -r -W -T contrib/apparmor/
+++ exec docker --daemon --debug --host unix:///root/gosrc/src/github.com/docker/docker/bundles/1.8.0-dev/test-integration-cli/docker.sock --storage-driver vfs --exec-driver native --pidfile bundles/1.8.0-dev/test-integration-cli/docker.pid --userland-proxy=true
---> Making bundle: .integration-daemon-setup (in bundles/1.8.0-dev/test-integration-cli)
---> Making bundle: .ensure-emptyfs (in bundles/1.8.0-dev/test-integration-cli)
---> Making bundle: .ensure-frozen-images (in bundles/1.8.0-dev/test-integration-cli)
---> Making bundle: .ensure-httpserver (in bundles/1.8.0-dev/test-integration-cli)
+ go test -check.f DockerSuite.TestCleanupMountsAfterCrash -test.timeout=60m -check.v github.com/docker/docker/integration-cli
OK: 0 passed
PASS
coverage: 1.6% of statements
ok      _/root/gosrc/src/github.com/docker/docker/integration-cli       0.030s
---> Making bundle: .integration-daemon-stop (in bundles/1.8.0-dev/test-integration-cli)
++++ cat bundles/1.8.0-dev/test-integration-cli/docker.pid
+++ kill 15342
+++ /etc/init.d/apparmor stop
Stopping apparmor (via systemctl): apparmor.service.
```

cpuguy83 commented:
@mrunalp No tests ran on that run.
See `0 passed`

mrunalp commented:
@cpuguy83 Am I invoking the command wrong?

cpuguy83 commented:
I actually don't see a `TestCleanupMountsAfterCrash` test in the PR.

cpuguy83 commented:
@mrunalp Ah yes, the problem is you are specifying `DockerSuite.TestCleanupMountsAfterCrash` where you really just want `TestCleanupMountsAfterCrash`

mrunalp commented:
@cpuguy83 Ah, okay, thanks :)

mrunalp commented:
Pushed an update which made the tests pass for me locally.

mrunalp commented:
Got past that particular test failure. Looking at the new ones.

mrunalp commented:
Alright, the tests are all green now :)

cpuguy83 commented on daemon/volumes_unix.go:
```diff
@@ -38,7 +38,7 @@ func copyOwnership(source, destination string) error {
 // calls Setup() on each. It also looks to see if is a network mount such as
 // /etc/resolv.conf, and if it is not, appends it to the array of mounts.
-func (container *Container) setupMounts() ([]execdriver.Mount, error) {
+func (container *Container) setupMounts(includeIpcMounts bool) ([]execdriver.Mount, error) {
```
Can we do this without the bool argument?
I don't think we need to worry about it since it's possible to copy this today via the API, which is what `mountVolumes` is used for.

mrunalp replied:
Updated.

cpuguy83 commented on daemon/volumes_windows.go:
```diff
@@ -18,7 +18,7 @@ func copyOwnership(source, destination string) error {
 // mount such as /etc/resolv.conf, and if it is not, appends it to the array
 // of mounts. As Windows does not support mount points, this is a no-op.
-func (container *Container) setupMounts() ([]execdriver.Mount, error) {
+func (container *Container) setupMounts(includeIpcMounts bool) ([]execdriver.Mount, error) {
```
Forgot one :(

mrunalp replied:
Fixed.

mrunalp committed:
Modify test to include /dev/shm sharing

Signed-off-by: Mrunal Patel <mrunalp@gmail.com>

mrunalp commented:
@cpuguy83 Added test changes.

mrunalp commented:
The windows build needs a bounce.

thaJeztah commented:
All green now. @cpuguy83 ptal :-)

cpuguy83 commented:
LGTM

thaJeztah commented:
ping @crosbymichael @icecrime think you were reviewing/testing this as well

cpuguy83 commented:
Needs a rebase.

mrunalp commented:
@cpuguy83 Rebased.

mrunalp commented:
Windows failures seem unrelated to this PR. Rest are green.

jfrazelle commented:
LGTM
